OCD-3621: Add/Edit Optional Standards

andlar opened the pull request.

andlar committed:
feat-flag!: Support add/edit/view of Optional Standards

[#OCD-3621]

andlar committed:
fix: show no optional standards when flag is off

[#OCD-3621]

andlar committed:
feat: put empty array in OS data when flag is off

[#OCD-3621]

andlar committed:
fix: save OS correctly on add/edit

[#OCD-3621]

andlar committed:
feat: add validation related to removing Test Standards

[#OCD-3621]

andlar committed:
feat: only show TS error/warning when OS flag is on

[#OCD-3621]

kekey1 commented on chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/SearchViewController.java:
```diff
@@ -994,6 +995,23 @@ private FuzzyChoices updateFuzzyChoices(FuzzyChoices fuzzyChoices)
 
+    @ApiOperation(value = "Get all possible optional standard options in the CHPL",
+            notes = "This is useful for knowing what values one might possibly search for.")
+    @RequestMapping(value = "/data/optional_standards", method = RequestMethod.GET,
```
should we use the dash notation here for the URL? /data/optional-standards

andlar committed:
feat: use mapping table for OS to criteria

[#OCD-3621]

andlar committed:
feat: provide "allowedOptionalStandards" in certification result

[#OCD-3621]

andlar committed:
feat: only show allowed OS when flag is on

[#OCD-3621]

kekey1 commented on chpl/chpl-service/src/main/java/gov/healthit/chpl/certifiedproduct/service/CertificationResultService.java:
```diff
@@ -31,23 +36,28 @@
 
     public List<CertificationResult> getCertificationResults(CertifiedProductSearchDetails searchDetails) throws EntityRetrievalException {
         List<SvapCriteriaMap> svapCriteriaMap = svapDao.getAllSvapCriteriaMap();
+        List<OptionalStandardCriteriaMap> optionalStandardCriteriaMap = optionalStandardDAO.getAllOptionalStandardCriteriaMap();
```
Something in the method name that's called here reads like it should be plural... getAllOptionalStandardCriteriaMaps()?

andlar replied:
I can see that, but I was following the pattern of the SVAP map in the line above. I feel like changing one of them would imply I should change the other, and I don't know how far that might ripple

kekey1 commented on chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/listing/reviewer/edition2015/OptionalStandardReviewer.java:
```diff
@@ -0,0 +1,48 @@
+    public void review(CertifiedProductSearchDetails listing) {
+        if (ff4j.check(FeatureList.OPTIONAL_STANDARDS)) {
+            listing.getCertificationResults().stream()
+            .filter(cert -> (cert.getTestStandards() != null && cert.getTestStandards().size() > 0))
```
I think I'm confused about this reviewer... I know optional standards are replacing test standards and there is currently some overlap, and I saw this work in your demo. But this is checking against testStandards only as far as I can tell. Where is the code putting optional standards in the testStandards array? I can't find it but am probably missing it. Should there also be validation against certResult.getOptionalStandards()? What about checking that the optional standard chosen is one of the ones allowed for the criteria (yes someone would have to use the API for this).

andlar replied:
This one was poorly named. I renamed it as a separate reviewer that we might be able to delete later, and added a reviewer specifically for Optional Standards (including some tests for it)

andlar committed:
fix: review Optional Standards for validity

also move reviewer that requires Test Standards to be removed to more
appropriate name

[#OCD-3621]

andlar committed:
test: add OptionalStandardReviewer tests

[#OCD-3621]

tmy1313 commented on chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/listing/reviewer/OptionalStandardReviewer.java:
```diff
@@ -0,0 +1,85 @@
+
+            Map<Long, List<OptionalStandardCriteriaMap>> optionalStandardCriteriaMap = null;
+            try {
+            optionalStandardCriteriaMap = optionalStandardDAO.getAllOptionalStandardCriteriaMap().stream()
```
Indent...

andlar committed:
fix: get edition for criteria

[#OCD-3621]

andlar committed:
feat: put Optional Standards in XML/XSD correctly

[#OCD-3621]

andlar committed:
fix: restrict OS existence when flag is off

[#OCD-3621]

andlar committed:
feat: add "citation" for Optional Standards

[#OCD-3621]

andlar committed:
docs: put Optional Standard description text in XSD

[#OCD-3621]

andlar committed:
feat: enable Optional Standards for 2014 Edition

[#OCD-3621]
